sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala

```diff
@@ -16,53 +16,168 @@
+          getClass.getClassLoader)
+      }
+      tmpClassLoader
+    } else {
+      pluginClassLoader
+    }
+  }
+
+  private def detectShimProvider(): String = {
     val sparkVersion = getSparkVersion
     logInfo(s"Loading shim for Spark version: $sparkVersion")
 
-    // This is not ideal, but pass the version in here because otherwise loader that match the
-    // same version (3.0.1 Apache and 3.0.1 Databricks) would need to know how to differentiate.
-    val sparkShimLoaders = ServiceLoader.load(classOf[SparkShimServiceProvider])
-        .asScala.filter(_.matchesVersion(sparkVersion))
-    if (sparkShimLoaders.size > 1) {
-      throw new IllegalArgumentException(s"Multiple Spark Shim Loaders found: $sparkShimLoaders")
+    // Emulating service loader manually because we have a non-standard jar layout for classes
+    val serviceProviderListPath = "/META-INF/services/" + classOf[SparkShimServiceProvider].getName
+    val serviceProviderList = Option(getClass.getResource(serviceProviderListPath)).toSeq
+      .map(scala.io.Source.fromURL)
+      .flatMap(_.getLines())
+
+    val shimServiceProviderOpt = serviceProviderList.flatMap { shimServiceProviderStr =>
+      val mask = shimIdFromPackageName(shimServiceProviderStr)
+      try {
+        val shimURL = new java.net.URL(s"${shimRootURL.toString}$mask/")
+        val shimClassLoader = new MutableURLClassLoader(Array(shimURL, shimCommonURL),
+          getClass.getClassLoader)
+        // can't use ServiceLoader with parallel world layout
+        val shimClass = shimClassLoader.loadClass(shimServiceProviderStr)
+        Option(
+          (instantiateClass(shimClass).asInstanceOf[SparkShimServiceProvider], shimURL)
+        )
+      } catch {
+        case cnf: ClassNotFoundException =>
+          logWarning(cnf + ": Could not load the provider", cnf)
+          None
+      }
+    }.find { case (shimServiceProvider, _) =>
+      shimServiceProvider.matchesVersion(sparkVersion)
+    }.map { case (inst, url) =>
+      shimURL = url
+      // this class will be loaded again by the real executor classloader
+      inst.getClass.getName
     }
-    logInfo(s"Found shims: $sparkShimLoaders")
-    val loader = sparkShimLoaders.headOption match {
-      case Some(loader) => loader
-      case None =>
+
+    shimServiceProviderOpt.getOrElse {
         throw new IllegalArgumentException(s"Could not find Spark Shim Loader for $sparkVersion")
     }
-    loader
   }
 
-  private def findShimProvider(): SparkShimServiceProvider = {
+  private def shimIdFromPackageName(shimServiceProvider: SparkShimServiceProvider) = {
+    shimServiceProvider.getClass.getPackage.toString.split('.').last
+  }
+
+  private def shimIdFromPackageName(shimServiceProviderStr: String) = {
+    shimServiceProviderStr.split('.').takeRight(2).head
+  }
+
+  private def findShimProvider(): String = {
+    // TODO restore support for shim provider override
```

Comment:
just make sure we do this because Cloudera uses regularly

Reply:
is there a typical classname they use here?

Reply:
--conf "spark.rapids.shims-provider-override=com.nvidia.spark.rapids.shims.spark311cdh.SparkShimServiceProvider"

Reply:
shimId being part of the package name even for an override is what I hoped for. If we start *requiring* this restriction going forward. it will be easy to accommodate after #3411